src/openassetio-core/include/openassetio/EntityReference.hpp

```diff
@@ -62,8 +114,34 @@ class EntityReference final {
 
 static_assert(std::is_move_constructible_v<EntityReference>);
 static_assert(std::is_move_assignable_v<EntityReference>);
+static_assert(std::is_copy_constructible_v<EntityReference>);
+static_assert(std::is_copy_assignable_v<EntityReference>);
 
 /// A list of entity references, used or batch-first functions.
 using EntityReferences = std::vector<EntityReference>;
 }  // namespace OPENASSETIO_CORE_ABI_VERSION
 }  // namespace openassetio
+
+namespace std {
+template <>
+struct hash<openassetio::EntityReference> {
+  std::size_t operator()(const openassetio::EntityReference& ref) const noexcept {
+    // Construct hash using common hashing function found
+    // across the literature, e.g. boost::hash_combine. I.e.
+    // seed ^= hasher(value) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+
+    // int(2^32 / phi) (where phi is the golden ratio).
+    constexpr std::size_t kInvPhi = 0x9e3779b9;
+    // Small coprime shift distances to spread out the bits.
+    constexpr std::size_t kLShift = 6;
+    constexpr std::size_t kRShift = 2;
+    // Seed incorporating type hash, assuming initial seed of 0
+    // (note: hash_code() is not constexpr in C++17).
+    static const std::size_t kSeed = typeid(openassetio::EntityReference).hash_code() + kInvPhi;
+    // Precompute rhs of hash equation.
+    static const std::size_t kBitMixer = kInvPhi + (kSeed << kLShift) + (kSeed >> kRShift);
+
+    return kSeed ^ (std::hash<openassetio::Str>{}(ref.toString()) + kBitMixer);
```

Comment:
As the comparison operators just work on the member string, should this hash not be the hash of that string? (in terms of consistency)

(I also assume, as can't see the code in this PR, that `ref.toString()` returns `ref.entityReferenceString_`?)

Reply:
> As the comparison operators just work on the member string, should this hash not be the hash of that string? (in terms of consistency)

As in, don't add the typeid to the hash?  

I think its consistent with the operators, in that you can't compare a string to an `EntityReference` directly (only if you explicitly call `.toString()`).  

I thought it a good idea to add typeid to the hash more for Python than for C++, where heterogeneous containers are a possibility (e.g. a `set` with both an `EntityReference` and a string). 

As it turns out, the C++ `std::hash` of a `std::string` gives a different value than the Python `__hash__` of a `str`, so despite there being a test adding both an `EntityReference` and a `str` to a container, there's fundamentally no chance of a clash. 

I did consider augmenting the tests to add a (test only) Python binding of `std::string` to show how they could coexist with `EntityReference`s in an associative container without clashing. However, it seemed like a lot of effort to test such a theoretical case.  Perhaps I should revisit that, if only to prove the point of including the typeid in the hash?

There's also an argument that the hash of an `EntityReference` may have other unforeseen uses where keeping those hashes distinct from the hash of a `std::string` would be useful. It just seems a best practice to disambiguate.

Happy to reconsider if there's a compelling reason?

> (I also assume, as can't see the code in this PR, that ref.toString() returns ref.entityReferenceString_?)

Yeah exactly.

Reply:
> I think its consistent with the operators, in that you can't compare a string to an EntityReference directly

I hadn't considered that - I was just thinking that if an Entity Reference _is_ a string (by composition), then maybe the hash should be the same.

Although the 'data' should be the same (which is what is in my head), the semantics of comparing two types should give a different value.

Resolving.